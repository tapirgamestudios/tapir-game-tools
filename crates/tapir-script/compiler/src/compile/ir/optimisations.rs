use std::{
    collections::HashMap,
    ops::{BitOr, BitOrAssign},
};

use petgraph::visit::Dfs;

use crate::{ast::SymbolId, compile::ir::TapIrFunction};

mod empty_phi;

#[derive(Clone, Copy, PartialEq, Eq)]
enum OptimisationResult {
    DidSomething,
    DidNothing,
}

impl BitOrAssign for OptimisationResult {
    fn bitor_assign(&mut self, rhs: Self) {
        *self = *self | rhs;
    }
}

impl BitOr for OptimisationResult {
    type Output = OptimisationResult;

    fn bitor(self, rhs: Self) -> Self::Output {
        if self == OptimisationResult::DidSomething || rhs == OptimisationResult::DidSomething {
            OptimisationResult::DidSomething
        } else {
            OptimisationResult::DidNothing
        }
    }
}

fn rename_all_variables(
    function: &mut TapIrFunction,
    renames: &HashMap<SymbolId, SymbolId>,
) -> OptimisationResult {
    if renames.is_empty() {
        return OptimisationResult::DidNothing;
    }

    let mut did_something = OptimisationResult::DidNothing;

    let mut dfs = Dfs::new(&*function, function.root);
    while let Some(block_id) = dfs.next(&*function) {
        let block = function.block_mut(block_id).expect("Could not find block");

        for symbol in block.sources_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }

        for symbol in block.targets_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }
    }

    did_something
}
