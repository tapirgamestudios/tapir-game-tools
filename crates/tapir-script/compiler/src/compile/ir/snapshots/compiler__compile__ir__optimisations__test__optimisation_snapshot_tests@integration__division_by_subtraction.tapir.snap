---
source: crates/tapir-script/compiler/src/compile/ir/optimisations.rs
expression: output
input_file: crates/tapir-script/compiler/src/compile/ir/snapshot_tests/optimisations/integration/division_by_subtraction.tapir
---
----------- before -------------
=======================
fn @toplevel() -> 

---- block 0 ----
temp.16 = int 100
getprop temp.17, 0
temp.18 = divide(temp.16, temp.17)
storeprop temp.18, 0
return 

=======================
fn divide(n.1, d.2) -> int

---- block 2 ----
r.19 = n.1
q.20 = int 0
jmp 0

---- block 0 ----
r.21 = φ(r.19 from 2, r.30 from 5)
d.23 = φ(d.2 from 2, d.23 from 5)
q.26 = φ(q.20 from 2, q.33 from 5)
temp.22 = r.21
temp.24 = d.23
temp.25 = temp.22 < temp.24
conditional_jump temp.25 3 4

---- block 3 ----
temp.27 = q.26
return temp.27

---- block 4 ----
jmp 5

---- block 5 ----
temp.28 = r.21
temp.29 = d.23
r.30 = temp.28 - temp.29
temp.31 = q.26
temp.32 = int 1
q.33 = temp.31 + temp.32
jmp 0


----------- optimised -------------
=======================
fn @toplevel() -> 

---- block 0 ----
temp.16 = int 100
getprop temp.17, 0
temp.18 = divide(temp.16, temp.17)
storeprop temp.18, 0
return 

=======================
fn divide(n.1, d.2) -> int

---- block 2 ----
q.20 = int 0
jmp 0

---- block 0 ----
r.21 = φ(n.1 from 2, r.30 from 5)
q.26 = φ(q.20 from 2, q.33 from 5)
temp.25 = r.21 < d.2
conditional_jump temp.25 3 5

---- block 3 ----
return q.26

---- block 5 ----
r.30 = r.21 - d.2
temp.32 = int 1
q.33 = q.26 + temp.32
jmp 0
